### PR TITLE
Support for generating replication task on workflow execution updates

### PR DESCRIPTION
**Pull request by samarabbas**
History engine changes to generate replication task when IsGlobalDomain
is enabled.  MutableStateBuilder is also updated to load replication
state for the domain and have support for creating replication task on
updates along with relevant updates to mutable state.

Replication tasks are written to a separate queue for the shard.  This
requires persistence changes to write replication tasks on updates in
the same transaction and new API for reading and completing the task.
New AckLevel is needed for processing of replication tasks.  This
requires changes to shard crud on the persistence layer for reading and
writing the new ack level for replication task processing.

ClusterMetadata and MessagingClient are already exposed on the Service.
So passing through Service to all layers in History service which needs
access to both ClusterMetadata and MessagingClient.

### CODE REVIEW

File: common/mocks/ClusterMetadata.go [hunk 1]
```diff
@@ -0,0 +1,114 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
```
Comment (wxing1292) [inline common/mocks/ClusterMetadata.go]: since metadataTestBase.go is provided, this mock does not do much, is it?

**Comment by wxing1292** [comment]
looks like we have a dedicated replication task queue, does that means each host will start additional # of shard of go routine for acking & processing replication task?

File: common/persistence/dataInterfaces.go [hunk 2]
```diff
@@ -73,7 +73,12 @@ const (
 	TransferTaskTypeCancelExecution
 	TransferTaskTypeStartChildExecution
 	TransferTaskTypeSignalExecution
-	TransferTaskTypeReplicationTask
+)
+
+// Types of replication tasks
+const (
+	ReplicationTaskTypeHistory = iota
+	ReplicationTaskTypeHeartbeat
```
Comment (wxing1292) [inline common/persistence/dataInterfaces.go]: is ReplicationTaskTypeHeartbeat in the next PR?

File: common/mocks/ClusterMetadata.go (hunk 1 above)
Comment (samarabbas) [reply common/mocks/ClusterMetadata.go]: I'm using it in some unit test to assert IsGlobalDomain API.

File: common/persistence/dataInterfaces.go (hunk 2 above)
Comment (samarabbas) [reply common/persistence/dataInterfaces.go]: Yes.  This is for future purpose.

**Pull request merged**